src/main/java/com/wavefront/service/CachedAppInfoFetcher.java

```diff
@@ -31,65 +36,69 @@
 @Component
 public class CachedAppInfoFetcher implements AppInfoFetcher {
 
-  private static final Logger logger = Logger.getLogger(CachedAppInfoFetcher.class.getCanonicalName());
-  /**
-   * We can tune this constant if needed ...
-   */
-  private static final int NUM_PCF_APPS = 5000;
+  private static final Logger logger = Logger.getLogger(
+      CachedAppInfoFetcher.class.getCanonicalName());
+
   /**
-   * AppData is mutable, so refresh the data every 1 hour
+   * How long to wait to receive AppInfo from PCF
    */
-  private static final int REFRESH_INTERVAL_HOUR = 1;
-  private final Cache<String, AppInfo> cache = CacheBuilder.newBuilder().
-          expireAfterWrite(REFRESH_INTERVAL_HOUR, TimeUnit.HOURS).maximumSize(NUM_PCF_APPS).build();
+  private static final int PCF_FETCH_TIMEOUT_SECONDS = 10;
+  private final Cache<String, Optional<AppInfo>> cache;
 
   @Autowired
   private CloudFoundryClient cfClient;
 
+  public CachedAppInfoFetcher(AppInfoProperties appInfoProperties) {
+    cache = Caffeine.newBuilder().
+        expireAfterWrite(appInfoProperties.getCacheExpireIntervalHours(), TimeUnit.HOURS).
+        maximumSize(appInfoProperties.getAppInfoCacheSize()).build();
+  }
+
   @Override
   public Mono<Optional<AppInfo>> fetch(String applicationId) {
-    // cache.get is thread safe
-    AppInfo appInfo = cache.getIfPresent(applicationId);
-    if (appInfo == null) {
+    // Note - cache.get is thread safe
+    Optional<AppInfo> optionalAppInfo = cache.getIfPresent(applicationId);
+    if (optionalAppInfo == null) {
```

Comment:
I think you might want to use `AsyncLoadingCache`, since it seems that you used get+put because of the asynchronous completion. Since the cache is `CompletableFuture` based, have the loader return `Mono#toFuture` and `Mono#fromFuture` from the cache.